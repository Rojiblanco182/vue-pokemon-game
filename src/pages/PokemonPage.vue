<template>
    <div v-if="!pokemon">
        <h3>Espere por favor...</h3>
    </div>

    <div v-else>
        <h1>Cuál es este pokémon?</h1>
        <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
        <PokemonOptions :pokemons="pokemonArr" @selection="showSelectedOption($event)" />
        
        <template v-if="showAnswer">
            <h2> {{ message }} </h2>
            <button @click="restart">Nuevo juego</button>
        </template>
    </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'
import getPokemonOptions from '@/helpers/getPokemonOptions'


export default {
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions()
            const randomNumber = Math.floor(Math.random() * 4)
            this.pokemon = this.pokemonArr[randomNumber]
        },
        showSelectedOption(selectedOption) {
            this.showPokemon = true
            this.showAnswer = true
            this.message = selectedOption === this.pokemon.id
            ? `Correcto, es ${this.pokemon.name}`
            : `Oops, fallaste! Es ${this.pokemon.name}`
        },
        restart() {
            this.showPokemon = false,
            this.showAnswer = false,
            this.message = '',
            this.pokemon = null,
            this.mixPokemonArray()
        }
    },
    mounted() {
        this.mixPokemonArray()
    }
}
</script>
