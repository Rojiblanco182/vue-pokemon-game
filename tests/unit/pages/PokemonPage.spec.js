import PokemonPage from '@/pages/PokemonPage'
import { shallowMount } from '@vue/test-utils'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage component', () => {
    test('should call mixPokemonArray() when the component is mounted', () => {
        const spyOnMixPokemonArray = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        shallowMount(PokemonPage)

        expect(spyOnMixPokemonArray).toHaveBeenCalled()
    })
    
    test('should match the snapshot once the data has been loaded', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''        
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })
    
    test('should show the PokemonOptions and PokemonPicture components', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''        
                }
            }
        })
        const pokemonPicture = wrapper.find('pokemon-picture-stub')
        const pokemonOptions = wrapper.find('pokemon-options-stub')

        expect(pokemonPicture.exists()).toBeTruthy()
        expect(Number(pokemonPicture.attributes('pokemonid'))).toBe(pokemons[0].id)
        expect(pokemonOptions.exists()).toBeTruthy()
        expect(pokemonOptions.attributes('pokemons')).toBeTruthy()
    })
    
    test('should work properly when showSelectedOption is called', async() => {
        const expectedWrongMessage = `Oops, fallaste! Es ${pokemons[0].name}`
        const expectedRightMessage = `Correcto, es ${pokemons[0].name}`
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''        
                }
            }
        })

        await wrapper.vm.showSelectedOption(pokemons[0].id)

        expect(wrapper.vm.message).toBe(expectedRightMessage)
        
        await wrapper.vm.showSelectedOption(8)
        
        expect(wrapper.vm.message).toBe(expectedWrongMessage)

        expect(wrapper.vm.showAnswer).toBe(true)
        expect(wrapper.vm.showPokemon).toBe(true)
    })
})
