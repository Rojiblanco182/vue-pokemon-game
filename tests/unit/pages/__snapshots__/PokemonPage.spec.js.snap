// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component should match the snapshot once the data has been loaded 1`] = `
<div>
  <h1>Cuál es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
